enable migrations to run within an external transaction

wenzowski opened the pull request:
After reading #2076 I'm not sure why we can't pass a transaction to `Migrator`?

wenzowski commented:
```javascript
const rollback = new Error('rollback');
await knex.transaction(async (trx) => {
  await knex.migrate.latest({trx});
  expect(true).toBe(true) // run some tests
  await knex.rollback({trx});
  throw rollback; // never actually commit
}).catch((error) => {
  if (error === rollback) return;
  console.error(error && error.stack);
  throw error;
});
```

elhigu commented:
This is missing at least tests and documentation. I'm not sure if this is necessary either. Wouldn't it be enough to just pass the transaction instead of knex pool to a constructor of the class? 

```
knex.transaction(async (trx) => {
  let migrator = new Migrator(trx);
  await migrator.xxxxx(yyyy);
});
```

wenzowski commented:
yes & yes; coming in future commit

I think a sane api would be something like
```javascript
knex.transaction(async (trx) => {
  trx.migrate.latest();
})
```
and the `trx` would stay out of `options`

wenzowski commented:
oh, and this seems absolutely crucial for transaction-based tests

```javascript
knex.transaction(async (trx) => {
  await knex.migrate.latest({trx});
  const conn = await trx.client.acquireConnection();
  // interact with a raw connection inside transaction context
  throw rollback;
});
```

elhigu commented:
@wenzowski I don't understand the second example, why you need to acquire connection from clients (I don't remember what that does exactly when called through trx.client... it might just return the same connection all over again that you already have and throwing in the end I suppose is just there if you like to force rollback).

About the earlier example does that currently work?

```
knex.transaction(async (trx) => {
  await trx.migrate.latest(); // await was missing
})
```

or this

```
knex.transaction(trx => trx.migrate.latest())
```

If those work, maybe this PR can be closed?

rapzo commented:
Ah! Was about to open a PR with the same proposal. Good one @wenzowski

wenzowski commented:
`knex.transaction(trx => trx.migrate.latest())` does not work. It was the first thing I tried ;)

Regarding the 2nd example, there are all sorts of times it can be useful to have access to the database connection itself. `knex.raw()` doesn't guarantee subsequent calls will be transmitted over the same connection, so if you've got a schema introspection library or an efficient dataloader or anything else that needs to run directly against the driver, that's how you'd get the current connection within the transaction scope.

elhigu commented:
Ah right, so that was workaround since method trx.raw doesn't exist for some reason... That is one another thing that should be fixed. Does trx.schema or trx.schema.raw exist and does they pass queries to the correct connection? Docs are lacking much info about this stuff and how to do this properly :(

rapzo commented:
Couldn't we provide the transaction in the constructor as well, via an optional second argument. It is kind of useful since the class is exported and can be extended for project needs?

Also, the `setConfig` needs to be public and use assign from lodash if everything gets transpiled and `Object.assign` is there already?

wenzowski commented:
Ok, I'll revisit this with something like
```javascript
export default class Migrator {

  constructor(knex, isTrx = false ) {
    this.knex = knex;
    this.isTrx = isTrx;
    this.config = this.setConfig(knex.client.config.migrations);
  }

  // ...
}
```
and
```javascript
knex.transaction(trx => {
  const migrate = new Migrator(trx, true);
  migrate.latest();
})
```
with `knex.transaction(trx => trx.migrate.latest())` to call `new Migrator` as above

elhigu commented:
I don't like the idea of passing the info if knex instance is transaction or not to the constructor. It would be better to add some property to transaction object to allow checking directly if that is the case (this `knex.isTransaction()` method would be beneficial for other uses too). 

However IMO correct fix for this would be to just make migration API to work through trx instance in other words to make `knex.transaction(trx => trx.migrate.latest())` to work.

Also I'm still wondering this part (I haven't got time to thest this):

> [Is it true that] ... method trx.raw doesn't exist for some reason ... Does trx.schema or trx.schema.raw exist and does they pass queries to the correct connection?

wenzowski commented:
`knex.transaction(trx => trx.raw())` does exist, ignore comment above about `acquireConnection()`. I still need that for my use case; others should probably use `trx.raw()` unless there's a good reason not to (eg. #285).

Rewrote still needs integration tests but docker takes too long to run so I'm going to let CI handle them.

elhigu commented:
Also mariadb tests did seem to fail.

elhigu commented on src/util/make-knex.js:
```diff
@@ -43,6 +43,10 @@ export default function makeKnex(client) {
       return client.transaction(container, config)
     },
 
+    isTransaction() {
```
This part needs documentation

wenzowski commented:
Sorry for the commit noise, but I'm going to keep pushing here and letting CI run the tests. The whole docker test suite is super cool, but it's taking so long to run I can't get any other work done so I need it offloaded to CI. I'm unit testing before committing (nice and fast), but not running the entire suite.

No idea on mariadb, but I'll look into it

zsszatmari commented:
Hi! Just a thumbs up for this, seems important to me.

kibertoad commented:
Depends on https://github.com/tgriesser/knex/pull/3084/files as it uses same simplified logic for determining what is a transaction.

elhigu commented:
Why I cant see any changes in tests now? 🤔

kibertoad commented:
@elhigu the ones that were there didn't test anything useful. How exactly you would test no second transaction is started?

elhigu commented:
> @elhigu the ones that were there didn't test anything useful. How exactly you would test no second transaction is started?

By mocking transaction start method and takin call count I suppose 🤔

elhigu commented:
or counting knex.on('query') events maybe

wenzowski commented:
I don't know if this PR is helpful in any way? I've long since abandoned use of `knex` in all my projects in favour of raw sql.

kibertoad commented:
I believe it's a valid case and don't see why not.

bentron2000 commented:
Just a note here that whilst the `isTransaction()` appears in the docs, the typescript defs do not include it. Cheers B

OlivierCavadenti committed:
feat: enable migrations to run within an external transaction

Reuse existing transaction in migrate methods instead of creating nested ones.
Enables transaction-based testing patterns (e.g. trx.migrate.latest()).

OlivierCavadenti committed:
docs: document running migrations within an external transaction

OlivierCavadenti committed:
fix: do not filter completed migrations when running down within a transaction

When migrate down/rollback runs inside an external transaction, _runBatch
received the trx and unconditionally filtered out already-completed
migrations via getNewMigrations. That filter is only correct for the "up"
direction; for "down" the supplied migrations ARE the completed ones to
undo, so filtering left an empty batch and nothing was rolled back.

Only resolve the completed list (and thus filter) when migrating up.

OlivierCavadenti committed:
test: roll external-transaction migration tests back via throw

Calling trx.rollback() explicitly and then letting the transaction
callback resolve makes Knex try to both roll back and commit the same
transaction, which hangs on MSSQL. Throwing a sentinel error rolls the
transaction back cleanly, so the latest()/rollback() cases now pass on
MSSQL too (verified against a local SQL Server and PostgreSQL).

Skip the rollback-dependent cases on engines that implicitly commit on
DDL and therefore cannot roll a migration back (MySQL/MariaDB, Oracle)
plus SQLite. Docs updated to match and to use the throw-based pattern.

OlivierCavadenti commented:
Hi !
I rebase and fixed up the tests and docs, CI is green now. 
@kibertoad @tgriesser could you take a look when you get a chance?
